### PR TITLE
DAOS-15040 pool: Disconnect handles when stopping (#13680)

### PR DESCRIPTION
**Pull request by wangdi1**
The ds_pool_stop function does not disconnect any associated ds_pool_hdl objects, which hold ds_pool references. When stopping an engine, we rely on the pool_stop_all function to delete all ds_pool_hdl objects. When destroying a pool, we rely, however, on a distributed pool evict operation invoked by the control plane, which may not work if the PS is unavailable or if the PS can't communicate with some engines.

To make ds_pool_stop and thus the pool destroy process more robust, this patch disconnects all associated ds_pool_hdl objects when stopping a pool.

  - Link all ds_pool_hdl objects onto a list in the ds_pool object.

  - Fix pool_iv_ent_refresh so that we can invalidate pool connections when the ds_pool object is stopping.

  - Remove the no-longer-necessary ds_pool_hdl_delete_all.

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by kccain — APPROVED** [review]

File: src/tests/suite/daos_mgmt.c [hunk 1]
```diff
@@ -440,6 +440,50 @@ get_sys_info_test(void **state)
...
+			     256 * 1024 * 1024 /* minimal size */, 0 /* nvme size */,
+			     NULL /* prop */, arg->pool.svc, uuid);
+	assert_rc_equal(rc, 0);
+	print_message("success uuid = "DF_UUIDF"\n", DP_UUID(uuid));
+
+	uuid_unparse_lower(uuid, uuid_str);
+	rc = daos_pool_connect(uuid_str, arg->group, DAOS_PC_RW, &arg->pool.poh,
+			       NULL /* pool info */, NULL /* ev */);
+
+	rc = daos_debug_set_params(arg->group, -1, DMG_KEY_FAIL_LOC,
```
Comment (kccain) [inline src/tests/suite/daos_mgmt.c:470]: LGTM.

For potential future expanded tests I wonder about another potential test
- pool connect fails after pool_connect_iv_dist() potentially leaving stranded ds_pool_tgt_hdl on storage engines (and leaving no record of the handle in the PS in rdb I believe is what we had seen in testing). Then during subsequent destroy, it is left to pool_tgt_disconnect_all() to clean up, else strand storage.

**Review by kccain — APPROVED** [review]
LGTM. Just a thought about potential new test in the future (that requires some more carful analysis on my part to know for sure if the scenario can be constructed with fault injection testing).

File: src/tests/suite/daos_mgmt.c (hunk 1 above)
Comment (liw) [reply src/tests/suite/daos_mgmt.c:470]: @kccain, that indeed sounds like a more realistic test than the one I came up with here. Thanks.

**Review by NiuYawei — APPROVED** [review]

**Comment by liw** [comment]
> Test stage Functional Hardware Medium completed with status FAILURE. https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13891/7/execution/node/1513/log

I think this is due to IB network instability on Wolf.

**Review by NiuYawei — APPROVED** [review]

**Comment by wangdi1** [comment]
all CI tests passed. @daos-stack/daos-gatekeeper

**Comment by daltonbohning** [comment]
> all CI tests passed. @daos-stack/daos-gatekeeper

A VM test failed so HW testing didn't even run
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-13921/2/pipeline

**Comment by kccain** [comment]
> > all CI tests passed. @daos-stack/daos-gatekeeper
> 
> A VM test failed so HW testing didn't even run https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-13921/2/pipeline

@daltonbohning is that a build link from a different PR? For this one I think the most recent builds are [build 7](https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-13891/7/pipeline) and [build 8](https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-13891/8/pipeline) that I think restarted from test hw stage.

**Comment by daltonbohning** [comment]
> > > all CI tests passed. @daos-stack/daos-gatekeeper
> > 
> > 
> > A VM test failed so HW testing didn't even run https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-13921/2/pipeline
> 
> @daltonbohning is that a build link from a different PR? For this one I think the most recent builds are [build 7](https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-13891/7/pipeline) and [build 8](https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-13891/8/pipeline) that I think restarted from test hw stage.

Wow, you're right. The Jenkins links below take me to the wrong PR!

**Review by gnailzenh — APPROVED** [review]

**Pull request merged**